File: 0207-course-schedule/0207-course-schedule.cpp
class Solution {
public:
    bool canFinish(int numCourses, vector<vector<int>>& prerequisites) {
        
        // build adjacency list
        vector<int> adj[numCourses];
        for(int i=0; i<prerequisites.size(); i++){
            adj[prerequisites[i][1]].push_back(prerequisites[i][0]);
        }
        
        // create indegree array
        vector<int> indegree(numCourses, 0);
        for(int i=0; i<numCourses; i++){
            for(auto node: adj[i]){
                indegree[node]++;
            }
        }
        
        queue<int> q;
        for(int i=0; i<numCourses; i++){
            if(indegree[i]==0) q.push(i);
        }
        
        int cnt=0;
        while(!q.empty()){
            int node = q.front();
            q.pop();
            cnt++;
            
            for(auto nbr: adj[node]){
                indegree[nbr]--;
                if(indegree[nbr]==0) q.push(nbr);
            }
        }
        
        return cnt == numCourses;
    }
};

File: 0207-course-schedule/0207-course-schedule.java
class Solution {
    
    boolean isDFSCycle(int u, List<List<Integer>> adj, boolean[] visited, boolean[] inRecursion){
        
        visited[u]=true;
        inRecursion[u]=true;
        
        for(int v : adj.get(u)){
            if(visited[v]==false && isDFSCycle(v, adj, visited, inRecursion)){
                return true;
            }
            else if(inRecursion[v] == true){
                return true;
            }
        }
        
        inRecursion[u]=false;
        return false;
    }
    
    boolean dfs(int numCourses, int[][] prerequisites) {
        
        boolean[] visited = new boolean[numCourses];
        boolean[] inRecursion = new boolean[numCourses];
        
        List<List<Integer>> adj = new ArrayList<>();
        for(int i=0;i<numCourses;i++){
            adj.add(new ArrayList<>());
        }
        
        for(int i=0; i<prerequisites.length; i++){
            int u = prerequisites[i][0];
            int v = prerequisites[i][1];
            
            adj.get(v).add(u);
        }
        
        for(int i=0; i<numCourses; i++){
            if(visited[i] == false && isDFSCycle(i, adj, visited, inRecursion)==true){
                return false;
            }
        }
        
        return true;
    }
    
    boolean topoSort(List<List<Integer>> adj, int[] indegree, int numCourses){
        
        Queue<Integer> q = new LinkedList<>();
        
        int count=0;
        for(int i=0; i<indegree.length; i++){
            if(indegree[i]==0){
                q.offer(i);
                count++;
            }
        }
        
        
        while(q.size()>0){
            
            int u = q.poll();
            
            for(int v: adj.get(u)){
                
                indegree[v]--;
                
                if(indegree[v] == 0){
                    q.offer(v);
                    count++;
                }
            }
        }
        
        return count == numCourses ? true : false;
    }
    
    public boolean canFinish(int numCourses, int[][] prerequisites) {
        
        List<List<Integer>> adj = new ArrayList<>();
        for(int i=0; i<numCourses; i++){
            adj.add(new ArrayList<>());
        }
        
        // create graph
        int[] indegree = new int[numCourses];
        for(int i=0; i<prerequisites.length; i++){
            int u = prerequisites[i][0];
            int v = prerequisites[i][1];
            
            adj.get(v).add(u); 
            indegree[u]++;
        }
        
        return topoSort(adj, indegree, numCourses);
    }
}